src/bootstrap/doc.rs

```diff
@@ -650,19 +651,20 @@ macro_rules! tool_doc {
                 let target = self.target;
                 builder.info(&format!("Documenting stage{} {} ({})", stage, stringify!($tool).to_lowercase(), target));
 
-                // This is the intended out directory for compiler documentation.
-                let out = builder.compiler_doc_out(target);
-                t!(fs::create_dir_all(&out));
-
-                let compiler = builder.compiler(stage, builder.config.build);
-
                 if !builder.config.compiler_docs {
                     builder.info("\tskipping - compiler/tool docs disabled");
                     return;
                 }
 
+                // This is the intended out directory for compiler documentation.
+                let out = builder.compiler_doc_out(target);
+                t!(fs::create_dir_all(&out));
+
                 // Build rustc docs so that we generate relative links.
                 builder.ensure(Rustc { stage, target });
+                // Rustdoc needs the rustc sysroot available to build.
+                let compiler = builder.compiler(stage, builder.config.build);
```

Comment:
Can we avoid putting the `let compiler` here? It looks like it's trying to imply this is somehow related to the check::Rustc just below, but that's not true.

Why is the check::Rustc necessary here? I would expect that the doc::Rustc just above ensures that the relevant metadata is available in the sysroot. Is the doc::Rustc step not writing/copying metadata into the sysroot? (I'd guess this is the case, but then it feels like maybe we should fix *that*...?)

check::Rustc doesn't seem like it should work in general, because it always writes into the stage0-sysroot (right?) but we might be documenting in stage 2 or stage 1...? There's also this comment which seems a little worrying - https://github.com/rust-lang/rust/blob/d2dfb0eb8e30d188fb1731e540bc1b418bcd046d/src/bootstrap/check.rs#L102-L103

Reply:
@Mark-Simulacrum so, I tried copying metadata into sysroot after documenting rustc:
```rust
diff --git a/src/bootstrap/doc.rs b/src/bootstrap/doc.rs
index d175796d77e..cfca5656717 100644
--- a/src/bootstrap/doc.rs
+++ b/src/bootstrap/doc.rs
@@ -13,6 +13,7 @@
 use std::path::{Path, PathBuf};
 
 use crate::Mode;
+use crate::compile::librustc_stamp;
 use build_helper::{t, up_to_date};
 
 use crate::builder::{Builder, Compiler, RunConfig, ShouldRun, Step};
@@ -615,7 +616,22 @@ fn run(self, builder: &Builder<'_>) {
             cargo.arg("-p").arg(krate);
         }
 
-        builder.run(&mut cargo.into());
+        compile::run_cargo(
+            builder,
+            cargo,
+            vec![],
+            &librustc_stamp(builder, compiler, target),
+            vec![],
+            true,
+        );
+
+        // Populate the sysroot (so that rustdoc can be documented without having to build the whole
+        // compiler, or play games with `check::Rustc`).
+        // We skip populating the sysroot in non-zero stage because that'll lead
+        // to rlib/rmeta conflicts if std gets built during this session.
+        let libdir = builder.sysroot_libdir(compiler, target);
+        let hostdir = builder.sysroot_libdir(compiler, compiler.host);
+        compile::add_to_sysroot(&builder, &libdir, &hostdir, &librustc_stamp(builder, compiler, target));
     }
 }
 
@@ -662,11 +678,9 @@ fn run(self, builder: &Builder<'_>) {
 
                 // Build rustc docs so that we generate relative links.
                 builder.ensure(Rustc { stage, target });
-                // Rustdoc needs the rustc sysroot available to build.
-                let compiler = builder.compiler(stage, builder.config.build);
-                builder.ensure(check::Rustc { target: builder.config.build });
 
                 // Symlink compiler docs to the output directory of rustdoc documentation.
+                let compiler = builder.compiler(stage, builder.config.build);
                 let out_dir = builder.stage_out(compiler, Mode::ToolRustc).join(target.triple).join("doc");
                 t!(fs::create_dir_all(&out_dir));
                 t!(symlink_dir_force(&builder.config, &out, &out_dir));
```
and it *almost* works, except that rustc_driver isn't checked before being documented, so it never ends up in the sysroot:
```
    Checking rustc_privacy v0.0.0 (/home/joshua/rustc/compiler/rustc_privacy)
 Documenting rustc_privacy v0.0.0 (/home/joshua/rustc/compiler/rustc_privacy)
    Checking rustc_interface v0.0.0 (/home/joshua/rustc/compiler/rustc_interface)
 Documenting rustc_interface v0.0.0 (/home/joshua/rustc/compiler/rustc_interface)
 Documenting rustc_driver v0.0.0 (/home/joshua/rustc/compiler/rustc_driver)
    Finished release [optimized + debuginfo] target(s) in 1m 16s
 Documenting rustdoc v0.0.0 (/home/joshua/rustc/src/librustdoc)
error[E0463]: can't find crate for `rustc_driver`
  --> src/librustdoc/lib.rs:38:1
   |
38 | extern crate rustc_driver;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate
```
I *think* I could hack this by running `cargo check -p rustc_driver` and then copying the resulting metadata into sysroot, but I'm not sure exactly how that would work ... I'd rather not copy too much of `run_cargo` if I don't have too.

Reply:
Yeah, no, this seems like a bad path to go down.

I'm surprised that the check::Rustc is enough though, per part of my previous comment here:

> check::Rustc doesn't seem like it should work in general, because it always writes into the stage0-sysroot (right?) but we might be documenting in stage 2 or stage 1...?

Have you checked this continues to work in all stages?

Reply:
Oh, huh, I forgot we changed check to not necessarily use the stage0 compiler all the time. OK then.